### PR TITLE
[expo-media-library] Fix compatibility with Android R

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Followup of #11647.

# How

Adjust expo-media-library module according to the scoped storage documentation.

> **Note:** Those changes were made with compatibility in mind. So not all changes are considered as a best practice on the newest Android. However, some of them may change the behavior of users app.

# TODO

- [x] Update docs

# Test Plan

(Tested on a simulator and real device with Android R)
- NCL ✅
- test-suite ✅

### CODE REVIEW

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MigrateAlbum.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  public MigrateAlbum(Context context, List<MediaLibraryUtils.AssetFile> assets, String albumDirName, Promise promise) {
+    mContext = context;
+    mAssets = assets;
+    mAlbumDirName = albumDirName;
+    mPromise = promise;
+  }
+
+  @Override
+  protected Void doInBackground(Void... voids) {
+    // Previously, users may save different assets type in the same directory.
```
Comment (tsapeta) [inline packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MigrateAlbum.java:40]:
```suggestion
    // Previously, users were able to save different assets type in the same directory.
```

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MigrateAlbum.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    // But now, it's not always possible.
+    // If album contains movies or pictures, we can move it to Environment.DIRECTORY_PICTURES.
+    // Otherwise, we reject.
+    Set<String> assetsRelativePaths = mAssets
+      .stream()
+      .map(asset -> getRelativePathForAssetType(asset.getMimeType(), false))
+      .collect(Collectors.toSet());
+
+    if (assetsRelativePaths.size() > 1) {
+      mPromise.reject(ERROR_UNABLE_TO_MIGRATE, "The album contains incompatible file types/");
```
Comment (tsapeta) [inline packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MigrateAlbum.java:50]:
```suggestion
      mPromise.reject(ERROR_UNABLE_TO_MIGRATE, "The album contains incompatible file types.");
```

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 3]
```diff
@@ -430,3 +430,20 @@ export async function getMomentsAsync() {
...
+// Android only
+export async function migrateAlbumAsync(album: AlbumRef): Promise<void> {
+  if (!MediaLibrary.migrateAlbumAsync) {
+    throw new UnavailabilityError('MediaLibrary', 'migrateAlbumAsync');
+  }
+
+  return await MediaLibrary.migrateAlbumAsync(getId(album));
+}
+
+export async function checkIfAlbumShouldBeMigratedAsync(album: AlbumRef): Promise<boolean> {
```
Comment (tsapeta) [inline packages/expo-media-library/src/MediaLibrary.ts:443]:
```suggestion
// Android only
export async function checkIfAlbumShouldBeMigratedAsync(album: AlbumRef): Promise<boolean> {
```

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 4]
```diff
@@ -430,3 +430,20 @@ export async function getMomentsAsync() {
 
   return await MediaLibrary.getMomentsAsync();
 }
+
+// Android only
+export async function migrateAlbumAsync(album: AlbumRef): Promise<void> {
+  if (!MediaLibrary.migrateAlbumAsync) {
+    throw new UnavailabilityError('MediaLibrary', 'migrateAlbumAsync');
+  }
```
Comment (tsapeta) [inline packages/expo-media-library/src/MediaLibrary.ts:438]: I'm wondering whether making it no-op on iOS wouldn't be better 🤔

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 5]
```diff
@@ -430,3 +430,20 @@ export async function getMomentsAsync() {
 
   return await MediaLibrary.getMomentsAsync();
 }
+
+// Android only
+export async function migrateAlbumAsync(album: AlbumRef): Promise<void> {
```
Comment (tsapeta) [inline packages/expo-media-library/src/MediaLibrary.ts:435]: I think we should clarify it's not always migrating, but only when it's needed 😉 Could you also add tsdocs for these new methods?

```suggestion
export async function migrateAlbumIfNeededAsync(album: AlbumRef): Promise<void> {
```

**Review by esamelson — COMMENTED** [review]
Wow, this required some hefty changes 💪 thanks for going through them all so thoroughly! Line-by-line the implementation looks good, I just have a couple of larger-scale questions/points to address:
- Is it possible for us to do some or all of these migrations automatically, or at least check automatically if they're needed (rather than exposing these TS functions)? It would be nice to minimize the amount of attention that developers need to pay to this change.
- Could you double check that the functional Java 8 features you've used will work here without issues? Our minApiVersion in Expo Go is 21 which AFAIK doesn't support Java 8 natively. (I think Android Studio will warn you about this if you open the Expo Go project.)

One other thing -- in the future, please put editor auto-formatting (indentation changes, moving anonymous listeners to lambdas, etc) into separate commits if you must include them at all. If they’re all mixed together, it makes the PR review longer & more error-prone because we have to sort through all the noise to find the meaningful changes. 🙏

**Comment by lukmccall** [comment]
> Is it possible for us to do some or all of these migrations automatically, or at least check automatically if they're needed (rather than exposing these TS functions)? It would be nice to minimize the amount of attention that developers need to pay to this change.

Unfortunately, no. Firstly, we can't assume that every album in the root directory was generated by us. Some of them may have been created by other apps - we don't want to migrate those. Moreover, some albums might contain various asset types which can be incompatible with each other. So we can't perform the migration.

> Could you double check that the functional Java 8 features you've used will work here without issues? Our minApiVersion in Expo Go is 21 which AFAIK doesn't support Java 8 natively. (I think Android Studio will warn you about this if you open the Expo Go project.)

Expo Go compiled for Android 21. 
Neither the `minSdkVersion` nor the `targetSdkVersion` is used to determine the source code compatibility for compilation (https://developer.android.com/studio/write/java8-support).

**Review by esamelson — COMMENTED** [review]
> Unfortunately, no. Firstly, we can't assume that every album in the root directory was generated by us. Some of them may have been created by other apps - we don't want to migrate those. Moreover, some albums might contain various asset types which can be incompatible with each other. So we can't perform the migration.

Gotcha. That's unfortunate, but not the first time we've had to deal with something annoying like this with an Android SDK upgrade. ☹️ Let's remember to call this out explicitly in the SDK 41 release notes (cc @brentvatne ).

One other question -- what happens if you try to use another MediaLibrary function on an album that needs to be migrated? Do we reject with an error that directs the developer to migrate the album?

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 6]
```diff
@@ -79,6 +79,43 @@ On **iOS 11+**, it's possible to use this method without asking for `CAMERA_ROLL
...
+
+- app is running on **iOS**, **web** or **Android below R**
+- app has **write permission** to the album folder
+
+The migration is possible when the album contains only compatible files types.
+For instance, movies and pictures are compatible with each other, but music and pictures are not.
+If automatic migration isn't possible, the function will be rejected.
+In that case, you can use methods from the `expo-file-system` to migrate all your files manually.
+
+#### Why do you need to migrate files?
```
[the remark this answers is not in the document]
Comment (lukmccall) [reply docs/pages/versions/unversioned/sdk/media-library.md:113]: I add note on the top of docs page which link to https://github.com/expo/fyi/pull/22 and migrate method

**Comment by lukmccall** [comment]
> One other question -- what happens if you try to use another MediaLibrary function on an album that needs to be migrated? Do we reject with an error that directs the developer to migrate the album?

I've added a better error message to `addAssetsToAlbum`. Other methods should work without any modifications.

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 7]
```diff
@@ -430,3 +430,43 @@ export async function getMomentsAsync() {
...
+// Android only
+/**
+ * Checks if provided album should be migrated.
+ * In other words, it checks if the application has the write permission to the album folder.
+ *
+ * This method always returns **false** for all android versions **below Android R**, **iOS** or **web**.
+ *
+ * @param album
+ */
+export async function checkIfAlbumShouldBeMigratedAsync(album: AlbumRef): Promise<boolean> {
```
Comment (esamelson) [inline packages/expo-media-library/src/MediaLibrary.ts:466]:
```suggestion
export async function albumNeedsMigrationAsync(album: AlbumRef): Promise<boolean> {
```

**Review by esamelson — APPROVED** [review]
one more change -- let's rename the check method to `albumNeedsMigrationAsync` to match the tense of the majority of our other boolean-returning methods (`isAvailableAsync`, `Location.hasServicesEnabledAsync`, etc.). Otherwise lgtm! 👍

**Pull request merged**